src/renderer/renderer.cpp

```diff
@@ -0,0 +1,45 @@
+#include "renderer.h"
+
+glm::vec3 Renderer::colorPixel(const unsigned int i, const unsigned int j,
+                               Scene &scene) {
+  // maps float from one range to another
+  auto map = [](float input, float oldMin, float oldMax, float newMin,
+                float newMax) {
+    float oldRange = oldMax - oldMin;
+    float newRange = newMax - newMin;
+    return ((input - oldMin) * newRange / oldRange) + newMin;
+  };
+
+  Camera &cam = scene.getCamera();
+  float fov = cam.getFOV();
+
+  float u = map(((float)i) + 0.5f, 0, mWidth, -1, 1);
+  float v = map(((float)j) + 0.5f, 0, mHeight, 1, -1);
+
+  float w = 1.0f / tanf(fov / 2.0f);
+
+  glm::vec3 p = u * cam.getRight() + v * cam.getUp() + w * cam.direction();
+
+  glm::vec3 rayOrg = cam.origin();
+  glm::vec3 rayDir = p - rayOrg;
```

Comment:
so that I make sure i'm following i,j are points on the whole screen we remap to unit space for u,v. with w you are getting the inverse tangent field of view. Now here is where i'm lost. p looks like a camera distance formula. I think it should be the destination point of the ray. thats why the subtraction from p by the origin gives us the direction.  I'm use to seeing p representing differently. So here is what I think I am seeing `cam.getRight()` and ` cam.getUp()` are so we start our orientation from the top right corner of the camera. u,v multiplication are so we get the offset  given i,j. and camera direction is so we know the right orientation. Some comments in this function would be helpful.

Reply:
I added some comments to better explain the process. If you still find it confusing we can jump on a call and discuss it and how to make the comments clearer.